File: cpp/cpp_100-999/188.cc
// C++
// leetcode 188
// https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-iv/

class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
    	vector<int> buy(k+1, INT_MIN);
    	vector<int> sell(k+1, 0);
    	int n = prices.size();
    	buy[0] = 0;
    	for(auto& price : prices) {
    		for(int i = 1; i <= k; i++) {
    			if(buy[i] == INT_MIN)
    				break;
    			sell[i] = max(sell[i], buy[i] + price); 
    		}
    		for(int i = 1; i <= k ; i++) {
    			buy[i] = max(buy[i], sell[i-1] - price);
    		}
    	}
    	return sell[k];
    }
};

File: Go/Go_100-999/188.go
// Go
// leetcode 188
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-iv/

func maxProfit(k int, prices []int) int {
	memo := make([][]int, k)
	for i:=0; i<k; i++ {
		memo[i] = []int{-100000,0}
	}
	for _, x := range prices {
		for i:=k-1; i> 0; i-- {
			memo[i][1] = max(memo[i][1], memo[i][0] + x)
			memo[i][0] = max(memo[i][0], memo[i-1][1] - x)
		}
		memo[0][1] = max(memo[0][1], memo[0][0] + x)
		memo[0][0] = max(memo[0][0], -x)
	}

	res := 0
	for i:=0; i<k; i++ {
		res = max(res, memo[i][1])
	}

	return res
}


func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}